From 4bba02c3eaed300eb6bf886d411ae330d278257a Mon Sep 17 00:00:00 2001
From: Nicolas LOUBOUTIN <nicolas.louboutin@st.com>
Date: Fri, 22 Jan 2021 09:12:48 +0100
Subject: [PATCH] usb: dwc2: reduce risk due to race condition

race between:
1- dwc2_hcd_init which will start a worker with dwc2_conn_id_status_change function
calling usb_gadget_map_request_by_dev which requires hsotg->gadget initialized
2- usb_add_gadget_udc which initialize hsotg->gadget


Signed-off-by: Nicolas LOUBOUTIN <nicolas.louboutin@st.com>
---
 drivers/usb/dwc2/platform.c | 19 ++++++++++---------
 1 file changed, 10 insertions(+), 9 deletions(-)

diff --git a/drivers/usb/dwc2/platform.c b/drivers/usb/dwc2/platform.c
index 0acbf58020ea..d7ef4b750a00 100644
--- a/drivers/usb/dwc2/platform.c
+++ b/drivers/usb/dwc2/platform.c
@@ -534,15 +534,6 @@ static int dwc2_driver_probe(struct platform_device *dev)
 		hsotg->hcd_enabled = 1;
 	}
 
-	platform_set_drvdata(dev, hsotg);
-	hsotg->hibernated = 0;
-
-	dwc2_debugfs_init(hsotg);
-
-	/* Gadget code manages lowlevel hw on its own */
-	if (hsotg->dr_mode == USB_DR_MODE_PERIPHERAL)
-		dwc2_lowlevel_hw_disable(hsotg);
-
 #if IS_ENABLED(CONFIG_USB_DWC2_PERIPHERAL) || \
 	IS_ENABLED(CONFIG_USB_DWC2_DUAL_ROLE)
 	/* Postponed adding a new gadget to the udc class driver list */
@@ -554,6 +545,16 @@ static int dwc2_driver_probe(struct platform_device *dev)
 		}
 	}
 #endif /* CONFIG_USB_DWC2_PERIPHERAL || CONFIG_USB_DWC2_DUAL_ROLE */
+
+	platform_set_drvdata(dev, hsotg);
+	hsotg->hibernated = 0;
+
+	dwc2_debugfs_init(hsotg);
+
+	/* Gadget code manages lowlevel hw on its own */
+	if (hsotg->dr_mode == USB_DR_MODE_PERIPHERAL)
+		dwc2_lowlevel_hw_disable(hsotg);
+
 	return 0;
 
 error_drd:
-- 
2.17.1

